Plutarch/Lift.hs

```diff
@@ -3,67 +3,77 @@
+type PLift = PUnsafeLiftDecl
 
-instance (Coercible h r, PLC.DefaultUni `PLC.Includes` r) => PUnsafeLiftDecl h (DerivePLiftViaCoercible h p r) where
-  type PLiftedRepr (DerivePLiftViaCoercible h p r) = r
-  type PLifted (DerivePLiftViaCoercible h p r) = h
-  pliftToRepr = coerce
-  pliftFromRepr = Just . coerce
-
-pconstant :: forall p h s. PUnsafeLiftDecl h p => h -> Term s p
-pconstant x = punsafeConstantInternal $ PLC.someValue @(PLiftedRepr p) @PLC.DefaultUni $ pliftToRepr @h @p x
+pconstant :: forall p s. PLift p => PLifted p -> Term s p
+pconstant x = punsafeConstantInternal $ PLC.someValue @(PConstantRepr (PLifted p)) @PLC.DefaultUni $ pconstantToRepr x
 
 -- | Error during script evaluation.
-data LiftError = LiftError deriving stock (Eq, Show)
+data LiftError
+  = LiftError_ScriptError Scripts.ScriptError
+  | LiftError_EvalException (ErrorWithCause (MachineError PLC.DefaultFun) ())
+  | LiftError_FromRepr
+  | LiftError_WrongRepr
+  deriving stock (Eq, Show)
 
-plift' :: forall p h. PUnsafeLiftDecl h p => ClosedTerm p -> Either LiftError h
+plift' :: forall p. PUnsafeLiftDecl p => ClosedTerm p -> Either LiftError (PLifted p)
 plift' prog = case evaluateScript (compile prog) of
   Right (_, _, Scripts.unScript -> UPLC.Program _ _ term) ->
-    case readKnownConstant @_ @(PLiftedRepr p) @(MachineError PLC.DefaultFun) Nothing term of
-      Right r -> case pliftFromRepr @h @p r of
+    case readKnownConstant @_ @(PConstantRepr (PLifted p)) @(MachineError PLC.DefaultFun) Nothing term of
+      Right r -> case pconstantFromRepr r of
         Just h -> Right h
-        Nothing -> Left LiftError
-      Left _ -> Left LiftError
-  Left _ -> Left LiftError
+        Nothing -> Left LiftError_FromRepr
+      Left e -> Left $ LiftError_EvalException e
+  Left e -> Left $ LiftError_ScriptError e
 
-plift :: forall p h. (HasCallStack, PUnsafeLiftDecl h p) => ClosedTerm p -> h
+plift :: forall p. (HasCallStack, PLift p) => ClosedTerm p -> (PLifted p)
 plift prog = case plift' prog of
   Right x -> x
-  Left _ -> error "plift failed"
+  Left e -> error $ "plift failed: " <> show e
+
+newtype DerivePConstantViaCoercible (h :: Type) (p :: PType) (r :: Type) = DerivePConstantViaCoercible h
+
+instance (PLift p, Coercible h r, PLC.DefaultUni `PLC.Includes` r) => PConstant (DerivePConstantViaCoercible h p r) where
+  type PConstantRepr (DerivePConstantViaCoercible h p r) = r
+  type PConstanted (DerivePConstantViaCoercible h p r) = p
+  pconstantToRepr = coerce
+  pconstantFromRepr = Just . coerce
+
+newtype DerivePConstantViaData (h :: Type) (p :: PType) = DerivePConstantViaData h
+
+instance (PLift p, Ledger.FromData h, Ledger.ToData h) => PConstant (DerivePConstantViaData h p) where
+  type PConstantRepr (DerivePConstantViaData h p) = Ledger.Data
+  type PConstanted (DerivePConstantViaData h p) = p
+  pconstantToRepr (DerivePConstantViaData x) = Ledger.toData x
+  pconstantFromRepr x = DerivePConstantViaData <$> Ledger.fromData x
```

Comment:
I'm not sure this leads to the correct handling under `PAsData`, I think it's only correct when `pfromData` and `pdata` are both id

Reply:
This isn't supposed to work with `PAsData` anyway, nor did it work with it.

Reply:
not sure if I've misunderstood the implications but I think it would mean a `Term s PFoo` from the result of `pfromData` vs from `pconstant` would be represented differently.